sdmx30-infomodel/src/main/java/com/epam/jsdmx/infomodel/sdmx30/VersionReference.java

```diff
@@ -1,58 +1,141 @@
+import static com.epam.jsdmx.infomodel.sdmx30.WildcardScope.MINOR;
+import static com.epam.jsdmx.infomodel.sdmx30.WildcardScope.NONE;
+
+import java.nio.channels.IllegalChannelGroupException;
 import java.util.Comparator;
-import java.util.List;
 import java.util.Objects;
-import java.util.function.Function;
-import java.util.regex.Pattern;
-
-import org.apache.commons.lang3.tuple.Pair;
 
 /**
  * Represents a wildcarded reference to a specific version of an artefact which is resolved according to the SDMX version management rules,
  * e.g. 1.2+.0
  */
 public final class VersionReference extends AbstractVersionReference {
 
-    private static final Pattern MAJOR_WILDCARD_PATTERN = Pattern.compile("(0|[1-9]\\d*)\\+\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(-\\w+)?");
-    private static final Pattern MINOR_WILDCARD_PATTERN = Pattern.compile("(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\+\\.(0|[1-9]\\d*)(-\\w+)?");
-    private static final Pattern PATCH_WILDCARD_PATTERN = Pattern.compile("(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\+(-\\w+)?");
-
-    private static final List<Pair<Pattern, Function<String, VersionReference>>> CREATORS_BY_PATTERN = List.of(
-        Pair.of(Version.PATTERN, (String s) -> new VersionReference(Version.createFromString(s), WildcardScope.NONE)),
-        Pair.of(MINOR_WILDCARD_PATTERN, (String s) -> new VersionReference(Version.createFromString(removeWildcard(s)), WildcardScope.MINOR)),
-        Pair.of(MAJOR_WILDCARD_PATTERN, (String s) -> new VersionReference(Version.createFromString(removeWildcard(s)), WildcardScope.MAJOR)),
-        Pair.of(PATCH_WILDCARD_PATTERN, (String s) -> new VersionReference(Version.createFromString(removeWildcard(s)), WildcardScope.PATCH))
-    );
-
     private VersionReference(Version version, WildcardScope scope) {
         super(version, scope);
     }
 
-    private static String removeWildcard(String s) {
-        return s.replaceAll("\\+", "");
-    }
-
     /**
      * Creates a {@link VersionReference} from a string representation.
+     *
      * @param version valid string representation of a referenced version
      */
     public static VersionReference createFromString(String version) {
-        return createFromString(CREATORS_BY_PATTERN, version);
+        try {
+            return parse(version);
+        } catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException("Invalid version: " + version);
+        }
+    }
+
+    private static VersionReference parse(String version) {
+        WildcardScope scope = NONE;
+        int majorUntil = -1;
+        int minorFrom = -1;
+        int minorUntil = -1;
+        int patchFrom = -1;
+        int patchUntil = -1;
+        int extensionFrom = -1;
+        Position p = Position.MAJOR;
+        final int inputLength = version.length();
+        for (int i = 0; i < inputLength; i++) {
+            char c = version.charAt(i);
+            if (c > 47 & c < 58) {
```

Comment:
47 and 58 should be constants, it's unclear write away that ASCII numbers start and end codes

Reply:
updated if to contain characters instead of ints, to ease the understandign